qiskit/algorithms/amplitude_amplifiers/grover.py

```diff
@@ -109,170 +108,123 @@ class Grover:
 
     """
 
-    @name_args([
-        ('oracle', ),
-        ('state_preparation', {bool: 'incremental'}),
-        ('iterations', ),
-        ('post_processing', {float: 'lam'}),
-        ('grover_operator', {list: 'rotation_counts'}),
-        ('quantum_instance', {str: 'mct_mode'}),
-        ('incremental', {(Backend, BaseBackend, QuantumInstance): 'quantum_instance'})
-    ], skip=1)  # skip the argument 'self'
     def __init__(self,
-                 oracle: Union[QuantumCircuit, Statevector],
-                 good_state: Optional[Union[Callable[[str], bool],
-                                            List[str], List[int], Statevector]] = None,
-                 state_preparation: Optional[QuantumCircuit] = None,
-                 iterations: Union[int, List[int]] = 1,
+                 iterations: Union[List[int], Iterator[int], float] = 1.2,
```

Comment:
Does this net out as the same default? One iteration I understood as the basic Grover.

Reply:
No it's not the same behaviour 🙂 Like this it tries powers 
```
1.2 ** 1, 1.2 ** 2, 1.2 ** 3, .... (casted to int)
```
This is suggested in some papers if the number of solutions is unknown and should somewhat reliably converge, which I don't think is the case for just using a single application of the Grover operator. I thought this might be a nicer default  behaviour, what do you think?

Reply:
As to the default I guess its a choice and the docstring is there with the explanation. Its just different than before and the basic Grover that perhaps one might expect. If this is a better default behavior for users of Grover its fine by mean, bear in mind people often just run things without reading too much and would what we had before be more the default expectation of the algorithm behavior?

Reply:
I discussed this with @stefan-woerner, in summary: if we know the number of solutions, then we should do a single iteration with the optimal power. In the more general case that we don't know the number of good states though, a growth rate is a more sensible default since it will likely still yield a correct solution. A single application of the grover operator however will not. 
So the suggestion is to 
* add a `num_good_states` or `num_solutions` to the `AmplificationProblem`
* leave iterations as `None` per default
* in `amplify` check if the iterations are None, and if yes: check if theres a number of solutions in the amplification problem, if yes compute the optimal power and do a single iterations, if no use a default growth rate

Reply:
I agree with it. We can use [`optimal_num_iterations`](https://github.com/Qiskit/qiskit-terra/pull/5680/files#diff-d9a4dc8032d5a9ebcd0d9c230509c077950d5d7614cab2bc06f0b4338a399cdfL276-R239) in `Grover` to calculate the optimal number of iterations.